File: package/server/app/schemas/location.py
from typing import Optional, List, Union
from uuid import UUID

from pydantic import BaseModel
from app.schemas.photo import Photo

class LocationBase(BaseModel):
    name: str
    level: str
    count: int

class LocationStatistics(BaseModel):
    province_count: int
    city_count: int
    district_count: int
    country_count: int

class Location(LocationBase):
    id: Optional[str] = None
    is_custom: Optional[bool] = None
    cover: Optional[Photo] = None

    class Config:
        from_attributes = True

class MapMarker(BaseModel):
    id: str
    lat: float
    lng: float

class LocationValue(BaseModel):
    province: Optional[str] = None
    city: Optional[str] = None
    district: Optional[str] = None

class LocationSearchItem(BaseModel):
    label: str
    value: LocationValue

class TimelineNode(BaseModel):
    type: str = "default"
    startDate: str
    endDate: str
    locationName: str
    lat: Optional[float] = None
    lng: Optional[float] = None
    photoCount: int = 0
    coverId: Optional[UUID] = None
    
class TimelineResponse(BaseModel):
    nodes: List[TimelineNode]
    total: int

File: package/website/src/types/location.ts
import type { Photo } from './album'

export interface Location {
  id?: string
  is_custom?: boolean
  name: string
  level: 'city' | 'province' | 'district' | 'scene'
  count: number
  cover: Photo | null
}

export interface LocationStatistics {
  province_count: number
  city_count: number
  district_count: number
  country_count: number
}

export interface Scene {
  id: string
  is_custom: boolean
  name: string
  description?: string
  level?: number
  address?: string
  latitude?: number
  longitude?: number
  radius?: number
  polygon?: number[][]
  photo_count?: number
  cover?: Photo | null
}

export interface SceneCreate {
  name: string
  description?: string
  level?: number
  address?: string
  latitude?: number
  longitude?: number
  radius?: number
  polygon?: number[][]
}

export interface SceneUpdate extends SceneCreate {}

export interface TimelineNode {
  type: string;
  startDate: string; // YYYY-MM-DD
  endDate: string; // YYYY-MM-DD
  locationName: string;
  lat?: number;
  lng?: number;
  photoCount: number;
  coverId?: string;
}

export interface TimelineResponse {
  nodes: TimelineNode[];
  total: number;
}
